lbrynet/core/Session.py

```diff
@@ -195,30 +198,54 @@ def _try_upnp(self):
 
         log.debug("In _try_upnp")
 
-        def get_free_port(upnp, port, protocol):
-            # returns an existing mapping if it exists
+        def get_mapped_port(upnp, port, protocol):
+            # returns an existing mapped port if it exists
             mapping = upnp.getspecificportmapping(port, protocol)
-            if not mapping:
-                return port
-            if upnp.lanaddr == mapping[0]:
-                return mapping
-            return get_free_port(upnp, port + 1, protocol)
-
-        def get_port_mapping(upnp, internal_port, protocol, description):
-            # try to map to the requested port, if there is already a mapping use the next external
-            # port available
+            if mapping is not None and upnp.lanaddr == mapping[0]:
+                return mapping[1]
+            return None
+
+        def add_port_mapping(upnp, port, auto_port, protocol, description):
+            log.debug('UPNP automatic port selection enabled: ' + str(auto_port))
+            try:
+                if auto_port is True:
+                    # addanyportmapping() returns None on failure
+                    return upnp.addanyportmapping(port, protocol, upnp.lanaddr, port,
```

Comment:
This fails with an ambiguous `UnknownError` if there is an existing redirect for another lan address, whereas `addportmapping` will raise `ConflictInMappingEntry`.

Reply:
weird, `addanyportmapping` is supposed to assign the first available port starting from the one specified as argument. Basically it is doing what we were doing before in our routine, but in an atomic and non-racy way. This is the main reason why I introduced it. It should then return the assigned port or none if it couldn't find a port.
But it sounds like you observed a different behaviour?